lib/telemetry_sampler.ex

```diff
@@ -1,5 +1,457 @@
 defmodule Telemetry.Sampler do
   @moduledoc """
-  Allows to periodically gather measurements and publish them as `Telemetry` events
+  Allows to periodically collect measurements and publish them as `Telemetry` events.
+
+  Measurements are MFAs called periodically by the Sampler process. These MFAs should collect
+  a value (if possible) and dispatch an event using `Telemetry.execute/3` function.
+
+  If the invokation of the MFA fails, the measurement is removed from the Sampler.
+
+  See the "Example - (...)" sections for more concrete examples.
+
+  ## Starting and stopping
+
+  You can start the Sampler using the `start_link/1` function. Sampler can be alaso started as a
+  part of your supervision tree, using both the old-style and the new-style child specifications:
+
+      # pre Elixir 1.5.0
+      children = [Supervisor.Spec.worker(Telemetry.Sampler, [[period: 5000]])]
+
+      # post Elixir 1.5.0
+      children = [{Telemetry.Sampler, [period: 5000]}]
+
+      Supervisor.start_link(children, [strategy: :one_for_one])
+
+  You can start as many Samplers as you wish, but generally you shouldn't need to do it, unless
+  you know that it's not keeping up with collecting all specified measurements.
+
+  Measurements need to be provided via `:measurements` option.
+
+  ## VM measurements
+
+  The `vm_measurements/1` function returns common measurements related to Erlang virtual machine
+  metrics. See its documentation for more information.
+
+  ## Example - measuring message queue length of the process
+
+  Measuring process' message queue length is a good way to find out if and when the process becomes
+  the bottleneck. If the length of the queue is growing, it means that the process is not keeping
+  up with the work it's been assigned and other processes asking it to do the work will get
+  timeouts. Let's try to simulate that situation using the following GenServer:
+
+      defmodule Worker do
+        use GenServer
+
+        def start_link() do
+          GenServer.start_link(__MODULE__, [])
+        end
+
+        def do_work(name) do
+          GenServer.call(name, :do_work, timeout = 5_000)
+        end
+
+        def init([]) do
+          {:ok, %{}}
+        end
+
+        def handle_call(:do_work, _, state) do
+          Process.sleep(1000)
+          {:reply, :ok, state}
+        end
+      end
+
+  When assigned with work (`handle_call/3`), the worker will sleep for 1 second to imitate long
+  running task.
+
+  Now we need a measurement dispatching the message queue length of the worker:
+
+      defmodule ExampleApp.Measurements do
+        def message_queue_length(pid) do
```

Comment:
Can we use a named process in this example? I think that's the most common case. Also, we need to match on the result of `Process.info`. So it would be something like:

```elixir
with pid when is_pid(pid) <- Process.whereis(name),
     {:message_queue_len, length} <- Process.info(pid, :message_queue_len) do
       Telemetry.execute([:example_app, :message_queue_length], length, %{pid: pid})
end
```